### DISCUSSION TITLE
请教关于address和nameserver规则优先级的问题

### DISCUSSION THREAD

**Discussion by Benjamin1919** [pymumu/smartdns#1588] [Q&A]
【Q1】当smartdns作为dnsmasq上游时，同时在系统hosts和smartdns配置中添加域名对应的IP，哪个优先级更高？
例如在系统hosts中添加：
`8.8.8.8 dns.google`
同时在smartdns配置中添加：
`address /dns.google/8.8.4.4`
请问这两条记录会合并吗？还是只应答其中的一个？

【Q2】我添加了两组上游DNS，国内组绑定6053端口，国外组绑定6553端口并且排除默认组，两组都没有设置跳过nameserver规则，然后我在配置中添加：
```
nameserver /domain-set:gfw-list/国外组
nameserver /domain-set:china-list/国内组
```
请问只访问6053端口时，gfw-list中的域名会通过国外组解析吗？
同理只访问6553端口时，china-list中的域名会通过国内组解析吗？

**Comment by PikuZheng** [comment]
1. 如果dnsmasq能处理hosts（我记得是能），它将不会向smartdns发起查询而是直接返回。
（如果没有dnsmasq，smartdns我记得是不处理hosts）
2. 不会（配置冲突），如果端口绑定了组，仅会向特定组的上游发起查询

**Reply by Benjamin1919** [reply]
感谢！如果6053端口没有绑定国内组，是不是就可以根据nameserver规则来使用国外组了呢？
另外此时国外组"-exclude-default-group"参数会不会导致6053端口不能使用国外组？

**Reply by PikuZheng** [reply]
> 感谢！如果6053端口没有绑定国内组，是不是就可以根据nameserver规则来使用国外组了呢？

正确

> 另外此时国外组"-exclude-default-group"参数会不会导致6053端口不能使用国外组？

这是个冲突配置，要么服务端口绑组，要么域名绑组，不要同时用。
如果6053端口没有绑组，从6053进来的查询会先去匹配nameserver规则，匹配到就走特定组，匹配不到的向所有非exclude-default-group上游查询（无论这个上游是什么组

**Reply by Benjamin1919** [reply]
不好意思，如果有空的话看一下这个视频：https://youtu.be/nkVO_MYXX1U?si=giqd4emKfygb1f4o&t=336
他的nameserver规则中用的是排除了默认组的上游服务器，而且他只开放了6053端口。第二服务器都还没启用，但是他可以正常解析。

我理解是不是即便是从默认组排除的服务器，6053端口也是可以根据nameserver规则来使用。只不过除了nameserver规则以外的就只能用没有从默认组排除的服务器了

**Reply by PikuZheng** [reply]
> 不好意思，如果有空的话看一下这个视频：https://youtu.be/nkVO_MYXX1U?si=giqd4emKfygb1f4o&t=336
> 他的nameserver规则中用的是排除了默认组的上游服务器，而且他只开放了6053端口。第二服务器都还没启用，但是他可以正常解析。

和我上面说的第二种方法有什么区别吗

> 我理解是不是即便是从默认组排除的服务器，6053端口也是可以根据nameserver规则来使用。只不过除了nameserver规则以外的就只能用没有从默认组排除的服务器了

这个问题我没看懂，“从默认组排除的服务器”和“6053端口根据nameserver规则来使用”逻辑上不存在关系；“除了nameserver规则以外”还有其他因素影响“是否使用从默认组排除的服务器”，不能一概而论。

smartdns发展至今，功能增增减减，配置会互相影响。有些配置组合逻辑上确实有漏洞，但也可以利用漏洞实现一些特殊需求。我建议是使用53端口，使用nameserver决定上游服务器，不要改不理解原理的配置。

**Reply by Benjamin1919** [reply]
哦哦不好意思我之前理解错了，我还以为排除默认组的服务器不能通过nameserver规则给6053端口用呢😂感谢解答
